### PR TITLE
Add timer functionality to allow timed events

### PR DESCRIPTION
**Pull request by nicolasgustafsson**
Also add constraint to StartAction

### CODE REVIEW

File: Game/GameWorld.cpp [hunk 1]
```diff
@@ -9,37 +9,55 @@
 
 #include "Designers/LevelDesigner/LevelDesigner.h"
 #include "Camera/CameraFunctionality.h"
+#include <Levels/LevelFunctionality.h>
```
Comment (Devilhar) [inline Game/GameWorld.cpp:12]: `#include "Levels/LevelFunctionality.h"`

File: Game/Levels/LevelFunctionality.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	
+	auto&& it = denizens.begin();
+	
+	while (it != denizens.end())
+	{
+		if (it->DenizenFunctionality.Get<SLevelDenizenComponent>().SlatedForRemoval)
+			it = denizens.erase(it);
+		else
+			it++;
+	}
```
Comment (Devilhar) [inline Game/Levels/LevelFunctionality.cpp:23]: Use remove_if erase idiom instead

File: Game/Levels/LevelFunctionality.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	for (auto&& object : aObjects)
+		AddDenizen(std::move(object));
+}
+
+void LevelFunctionality::TransferToNewLevel(LevelFunctionality& aNewLevel)
+{
+	for (auto&& denizen : Get<SLevelComponent>().Denizens)
+	{
+		if (denizen.DenizenFunctionality.Get<SLevelDenizenComponent>().PersistentOnLevelChange)
+			aNewLevel.AddDenizen(std::move(denizen.DenizenObject));
```
Comment (Devilhar) [inline Game/Levels/LevelFunctionality.cpp:63]: This makes the LevelFunctionality enter an invalid state. Please reset the functionality after this function if possible by calling: `Get<SLevelComponent>().Denizens.clear();`

File: Game/Levels/LevelFunctionality.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+	auto it = std::find_if(denizens.begin(), denizens.end(), [objectPtr](SLevelComponent::SDenizen& aDenizen) {return &aDenizen.DenizenObject == objectPtr; });
+
+	if (it != denizens.end())
+		denizens.erase(it);
+}
+
+void LevelFunctionality::AddDenizens(plf::colony<Object>&& aObjects)
+{
+	for (auto&& object : aObjects)
+		AddDenizen(std::move(object));
```
Comment (Devilhar) [inline Game/Levels/LevelFunctionality.cpp:55]: clear aObjects when done.

File: Game/Player/PlayerControllerFunctionality.cpp [hunk 5]
```diff
@@ -2,6 +2,9 @@
 #include "PlayerControllerFunctionality.h"
 #include "Weapons/MeleeWeapon.h"
 #include "Designers/MeleeWeaponDesigner/MeleeWeaponDesigner.h"
+#include <Enemy/EnemyControllerFunctionality.h>
+#include <UtilityFunctionalities/TimeToLiveFunctionality.h>
+#include <Levels/LevelFunctionality.h>
```
Comment (Devilhar) [inline Game/Player/PlayerControllerFunctionality.cpp:7]: Use "" instead of <>

File: Game/Player/PlayerControllerFunctionality.cpp [hunk 6]
```diff
@@ -64,6 +84,9 @@ void PlayerControllerFunctionality::OnDeath()
 	Get<SpriteRenderingFunctionality>().SetTexture("Textures/deadPlayer.png");
 
 	WmDispatchMessage(SPlayerDiedMessage());
+
+	Get<TimerFunctionality>().AddTimer(3.f, []() {WmLog("wao!"); });
+	Get<TimerFunctionality>().AddTimer(5.f, [&]() {WmLog("do it again!"); 	Get<TimerFunctionality>().AddTimer(5.f, [&]() {WmLog("and again!"); }); });
```
Comment (Devilhar) [inline Game/Player/PlayerControllerFunctionality.cpp:89]: Remove

File: Game/Player/PlayerControllerFunctionality.h [hunk 7]
```diff
@@ -12,6 +12,8 @@
 #include "Functionalities/TransformFunctionality.h"
 #include "Functionalities/OwnerFunctionality.h"
 #include "Functionalities/ImpulseFunctionality.h"
+#include <Functionalities/TimerFunctionality.h>
+#include <Levels/LevelDenizenFunctionality.h>
```
Comment (Devilhar) [inline Game/Player/PlayerControllerFunctionality.h:16]: Fix

File: Game/UtilityFunctionalities/FactionFunctionality.h [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+enum class EFaction
+{
+	Player,
+	Enemy,
+	Neutral
+};
+
+struct SFactionComponent : public SComponent
+{
+	EFaction Faction;
```
Comment (Devilhar) [inline Game/UtilityFunctionalities/FactionFunctionality.h:12]: Set default value

File: WonderMake/Collision/CollisionSystem.cpp [hunk 9]
```diff
@@ -137,7 +137,18 @@ void CollisionSystem::OverlapAgainstFunctionalityInternal(const Colliders::Shape
 
 bool CollisionSystem::TestSphereVsLineCollision(const Colliders::SSphere& aSphereA, const Colliders::SLine& aLineB) noexcept
 {
-	const SVector2f closestPoint = GetClosestPointOnLine(aLineB, aSphereA.Position);
+	SVector2f closestPoint = GetClosestPointOnLine(aLineB, aSphereA.Position);
```
Comment (Devilhar) [inline WonderMake/Collision/CollisionSystem.cpp:140]: const

File: WonderMake/Functionalities/BaseFunctionality.cpp [hunk 10]
```diff
@@ -0,0 +1,2 @@
+#include "pch.h"
+#include "BaseFunctionality.h"
```
Comment (Devilhar) [inline WonderMake/Functionalities/BaseFunctionality.cpp:2]: Remove file if not intended to be used.

File: WonderMake/Functionalities/TimerFunctionality.cpp [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+void TimerFunctionality::Tick() 
+{
+	//[Nicos]: We use two loops in this function to protect against adding timers in callbacks
+
+	STimerComponent& timerComponent = Get<STimerComponent>();
+
+	const f32 deltaTime = SystemPtr<TimeKeeper>()->GetDeltaSeconds();
+
+	for (auto&& timer : timerComponent.Timers)
```
Comment (Devilhar) [inline WonderMake/Functionalities/TimerFunctionality.cpp:24]: Change to `auto&`

File: WonderMake/Functionalities/TimerFunctionality.cpp [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+	while (it != timerComponent.Timers.end())
+	{
+		auto&& timer = *it;
+
+		if (timer.TimeLeft < 0.f)
+			it = timerComponent.Timers.erase(it);
+		else
+			it++;
+	}
```
Comment (Devilhar) [inline WonderMake/Functionalities/TimerFunctionality.cpp:42]: remove_if erase idiom

File: WonderMake/Graphics/Renderer.h [hunk 13]
```diff
@@ -40,6 +40,6 @@ class Renderer
 
 	bool myDebugWindowHasFocus = false;
 
-	const SColor ClearColor = SColor::CornflowerBlue;
+	const SColor ClearColor = SColor::Grey;
```
Comment (Devilhar) [inline WonderMake/Graphics/Renderer.h:43]: :(

File: WonderMake/Graphics/Shader.h [hunk 14]
```diff
@@ -21,16 +22,17 @@ class Shader
 	, private NonCopyable
 {
 public:
-	Shader(const std::filesystem::path Path)
+	Shader(const std::filesystem::path aPath)
```
Comment (Devilhar) [inline WonderMake/Graphics/Shader.h:25]: const ref

File: WonderMake/Logging/ImGuiLogger.cpp [hunk 15]
```diff
@@ -18,28 +19,47 @@ void ImGuiLogger::Debug()
 
 	ImGui::PushStyleVar(ImGuiStyleVar_ItemSpacing, ImVec2(4, 1)); // Tighten spacing
 
-	for (auto&& LogMessage : myLogMessages)
+
+	for (auto&& logMessage : myLogMessages)
```
Comment (Devilhar) [inline WonderMake/Logging/ImGuiLogger.cpp:23]: auto ref

File: WonderMake/Resources/ResourceSystem.h [hunk 16]
```diff
@@ -26,13 +28,56 @@ class ResourceSystem
...
+	MessageSubscriber mySubscriber;
 	std::mutex myLock;
 	std::unordered_map<std::string, std::weak_ptr<SResource<TResource>>> myResources;
 	std::unordered_map<std::string, std::shared_ptr<CreateResource<TResource>>> myCreateResourceJobs;
 };
 
+template<typename TResource>
+void ResourceSystem<TResource>::OnFileChange(const SFileChangedMessage& aFileChangedMessage)
+{
+	for (auto&& str : myResources)
```
Comment (Devilhar) [inline WonderMake/Resources/ResourceSystem.h:44]: auto ref

File: WonderMake/Utilities/Math.h [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+				return aMax;
+
+			return aValue;
+		}
+
+	//template <class T> requires std::totally_ordered_with<T, f32>
+	//T Sign(T)
+	//{
+	//	return aStart + ((aEnd - aStart) * aProgress);
+	//}
```
Comment (Devilhar) [inline WonderMake/Utilities/Math.h:38]: Remove

File: WonderMake/Functionalities/TimerFunctionality.cpp (hunk 11 above)
Comment (nicolasgustafsson) [reply WonderMake/Functionalities/TimerFunctionality.cpp:24]: Intended

File: WonderMake/Functionalities/TimerFunctionality.cpp (hunk 12 above)
Comment (nicolasgustafsson) [reply WonderMake/Functionalities/TimerFunctionality.cpp:42]: colony

File: WonderMake/Collision/CollisionSystem.cpp (hunk 9 above)
Comment (nicolasgustafsson) [reply WonderMake/Collision/CollisionSystem.cpp:140]: I modify it later in the function :)

File: Game/Levels/LevelFunctionality.cpp (hunk 3 above)
Comment (nicolasgustafsson) [reply Game/Levels/LevelFunctionality.cpp:66]: Fixed

File: Game/GameWorld.cpp (hunk 1 above)
Comment (nicolasgustafsson) [reply Game/GameWorld.cpp]: Fixed

File: Game/Levels/LevelFunctionality.cpp (hunk 4 above)
Comment (nicolasgustafsson) [reply Game/Levels/LevelFunctionality.cpp:56]: Fixed

File: Game/Player/PlayerControllerFunctionality.cpp (hunk 5 above)
Comment (nicolasgustafsson) [reply Game/Player/PlayerControllerFunctionality.cpp]: Fixed

File: Game/Player/PlayerControllerFunctionality.cpp (hunk 6 above)
Comment (nicolasgustafsson) [reply Game/Player/PlayerControllerFunctionality.cpp]: Fixed

File: Game/Player/PlayerControllerFunctionality.h (hunk 7 above)
Comment (nicolasgustafsson) [reply Game/Player/PlayerControllerFunctionality.h]: Fixed

File: Game/UtilityFunctionalities/FactionFunctionality.h (hunk 8 above)
Comment (nicolasgustafsson) [reply Game/UtilityFunctionalities/FactionFunctionality.h]: Fixed

File: WonderMake/Functionalities/BaseFunctionality.cpp (hunk 10 above)
Comment (nicolasgustafsson) [reply WonderMake/Functionalities/BaseFunctionality.cpp]: Fix'd

File: Game/Levels/LevelFunctionality.cpp (hunk 3 above)
Comment (Devilhar) [reply Game/Levels/LevelFunctionality.cpp:66]: Not seeing the fix

Comment (nicolasgustafsson) [reply Game/Levels/LevelFunctionality.cpp:65]: Fixed for realsies

**Pull request merged**